pepper-apis/src/main/java/org/broadinstitute/ddp/util/FilterUtil.java

```diff
@@ -0,0 +1,50 @@
+package org.broadinstitute.ddp.util;
+
+import static java.util.stream.Collectors.toList;
+import static spark.Spark.before;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.broadinstitute.ddp.filter.ExcludePathFilterWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.Filter;
+
+public class FilterUtil {
+    private static final Logger LOG = LoggerFactory.getLogger(FilterUtil.class);
+
+    /**
+     * Creates "before" filter on specified that checks that IP address matches one of those provided
+     * Filter will return a 404 and a page not found API error message
+     * @param path the URL path
+     * @param ips Collection of IPs to allow
+     */
+    public static void whitelist(String path, Collection<String> ips) {
+        Set<String> ipSet = (ips instanceof Set) ? (Set<String>) ips : new HashSet<>(ips);
+        before(path, (request, response) -> {
+            if (!(ipSet.contains(request.ip()))) {
```

Comment:
I think this needs to include something like `request.getHeader("X-Forwarded-For")`.  Docker backend is deployed behind a loadbalancer, and probably GAE has something similar?

Reply:
Looks like this subject is a little more complicated than my naive code! Digging deeper.

Reply:
OK. Wrote a little program and ran it in GAE. Below is the output when I do I GET from my home laptop.
Note that Google reports my laptops IP to be: `146.115.144.117`

**Your header info was:**
```
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9,es-US;q=0.8,es;q=0.7
Forwarded: for="146.115.144.117";proto=https
Host: headergrabber-dot-broad-ddp-dev.uc.r.appspot.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.163 Safari/537.36
X-AppEngine-City: newton
X-AppEngine-CityLatLong: 42.337041,-71.209221
X-AppEngine-Country: US
X-AppEngine-Default-Version-Hostname: broad-ddp-dev.appspot.com
X-AppEngine-Https: on
X-AppEngine-Region: ma
X-AppEngine-Request-Log-Id: 5e8dc6f900ff0ef01e62a019490001737e62726f61642d6464702d6465760001686561646572677261626265723a6c6f63616c74657374000100
X-AppEngine-User-IP: 146.115.144.117
X-Cloud-Trace-Context: 72314e2aed053d401218f63beb9dfe76/3843537310155855511
X-Forwarded-For: 146.115.144.117, 169.254.1.1
X-Forwarded-Proto: https
dnt: 1
sec-fetch-dest: document
sec-fetch-mode: navigate
sec-fetch-site: none
sec-fetch-user: ?1
upgrade-insecure-requests: 1
```
**Other Info:**
```
request.ip(): 146.115.144.117
request.raw().getRemoteAddr():146.115.144.117
request.raw().getRemoteHost():146.115.144.117
```

Reply:
**Going through Broad VPN** Google reports: `69.173.127.24`

**Your header info was:**
```
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9,es-US;q=0.8,es;q=0.7
Cache-Control: max-age=0
Forwarded: for="69.173.127.24";proto=https
Host: headergrabber-dot-broad-ddp-dev.uc.r.appspot.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.163 Safari/537.36
X-AppEngine-City: boston
X-AppEngine-CityLatLong: 42.360082,-71.058880
X-AppEngine-Country: US
X-AppEngine-Default-Version-Hostname: broad-ddp-dev.appspot.com
X-AppEngine-Https: on
X-AppEngine-Region: ma
X-AppEngine-Request-Log-Id: 5e8dca3a00ff0e5de9b5269db90001737e62726f61642d6464702d6465760001686561646572677261626265723a6c6f63616c74657374000100
X-AppEngine-User-IP: 69.173.127.24
X-Cloud-Trace-Context: 49d0026b60123328d4027a6b0b882722/11734570927955475461;o=1
X-Forwarded-For: 69.173.127.24, 169.254.1.1
X-Forwarded-Proto: https
dnt: 1
sec-fetch-dest: document
sec-fetch-mode: navigate
sec-fetch-site: none
sec-fetch-user: ?1
upgrade-insecure-requests: 1
```
**Other Info:**
```
request.ip(): 69.173.127.24
request.raw().getRemoteAddr():69.173.127.24
request.raw().getRemoteHost():69.173.127.24
```

Reply:
@zyme So appears that, at least running within GAE request.ip() is reporting accurately the host remote address (at least as accurately as it is possible in a proxy/gateway/blah/blah environment). `Matches X-Forward-For` header.
I suppose this is very dependent on the host environment and would behave differently if fronting our app with nginx or Apache.
So probably the generic solution would be to make this ip detection configurable so it can be set up differently in different deployment environments.
Or we can leave it as is if we are targeting GAE?

Reply:
By the way, if you want to try it yourself:

Reply:
Woo hoo!  So glad that GAE infrastructure gives the right answer here.